File: android/src/main/java/com/amplitude/android/internal/locators/ComposeViewTargetLocator.java
package com.amplitude.android.internal.locators;

import androidx.annotation.OptIn;
import androidx.compose.ui.InternalComposeUiApi;
import androidx.compose.ui.geometry.Rect;
import androidx.compose.ui.layout.ModifierInfo;
import androidx.compose.ui.node.LayoutNode;
import androidx.compose.ui.node.Owner;
import androidx.compose.ui.semantics.SemanticsConfiguration;
import androidx.compose.ui.semantics.SemanticsModifier;
import androidx.compose.ui.semantics.SemanticsPropertyKey;

import com.amplitude.android.internal.ViewTarget;
import com.amplitude.common.Logger;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayDeque;
import java.util.List;
import java.util.Map;
import java.util.Queue;

import kotlin.Pair;

@SuppressWarnings("KotlinInternalInJava")
@OptIn(markerClass = InternalComposeUiApi.class)
public class ComposeViewTargetLocator implements ViewTargetLocator {
    private volatile @Nullable ComposeLayoutNodeBoundsHelper composeLayoutNodeBoundsHelper;
    private static final String SOURCE = "jetpack_compose";
    private final @NotNull Logger logger;

    public ComposeViewTargetLocator(@NotNull Logger logger) {
        this.logger = logger;
    }

    @Nullable
    @Override
    public ViewTarget locate(
            @NotNull Object root,
            @NotNull Pair<Float, Float> targetPosition,
            @NotNull ViewTarget.Type targetType) {

        // lazy init composeHelper as it's using some reflection under the hood
        if (composeLayoutNodeBoundsHelper == null) {
            synchronized (this) {
                if (composeLayoutNodeBoundsHelper == null) {
                    composeLayoutNodeBoundsHelper = new ComposeLayoutNodeBoundsHelper(logger);
                }
            }
        }

        if (!(root instanceof Owner)) {
            return null;
        }

        final @NotNull Queue<LayoutNode> queue = new ArrayDeque<>();
        queue.add(((Owner) root).getRoot());

        // the final tag to return
        @Nullable String targetTag = null;

        // the last known tag when iterating the node tree
        @Nullable String lastKnownTag = null;
        while (!queue.isEmpty()) {
            final @Nullable LayoutNode node = queue.poll();
            if (node == null) {
                continue;
            }

            if (node.isPlaced() && layoutNodeBoundsContain(composeLayoutNodeBoundsHelper, node, targetPosition.component1(), targetPosition.component2())) {
                boolean isClickable = false;
                final List<ModifierInfo> modifiers = node.getModifierInfo();
                for (ModifierInfo modifierInfo : modifiers) {
                    if (modifierInfo.getModifier() instanceof SemanticsModifier) {
                        final SemanticsModifier semanticsModifierCore =
                                (SemanticsModifier) modifierInfo.getModifier();
                        final SemanticsConfiguration semanticsConfiguration =
                                semanticsModifierCore.getSemanticsConfiguration();
                        for (Map.Entry<? extends SemanticsPropertyKey<?>, ?> entry : semanticsConfiguration) {
                            final @Nullable String key = entry.getKey().getName();
                            if ("OnClick".equals(key)) {
                                isClickable = true;
                            } else if ("TestTag".equals(key)) {
                                if (entry.getValue() instanceof String) {
                                    lastKnownTag = (String) entry.getValue();
                                }
                            }
                        }
                    } else {
                        // Newer Jetpack Compose 1.5 uses Node modifiers for clicks/scrolls
                        final @Nullable String type = modifierInfo.getModifier().getClass().getCanonicalName();
                        if ("androidx.compose.foundation.ClickableElement".equals(type)
                                || "androidx.compose.foundation.CombinedClickableElement".equals(type)) {
                            isClickable = true;
                        }
                    }
                }

                if (isClickable && targetType == ViewTarget.Type.Clickable) {
                    targetTag = lastKnownTag;
                }
            }
            queue.addAll(node.getZSortedChildren().asMutableList());
        }

        if (targetTag == null) {
            return null;
        } else {
            return new ViewTarget(null, null, null, targetTag, SOURCE);
        }
    }

    private static boolean layoutNodeBoundsContain(
            @NotNull ComposeLayoutNodeBoundsHelper composeLayoutNodeBoundsHelper,
            @NotNull LayoutNode node,
            final float x,
            final float y) {

        final @Nullable Rect bounds = composeLayoutNodeBoundsHelper.getLayoutNodeWindowBounds(node);
        if (bounds == null) {
            return false;
        } else {
            return x >= bounds.getLeft()
                    && x <= bounds.getRight()
                    && y >= bounds.getTop()
                    && y <= bounds.getBottom();
        }
    }
}

File: android/src/main/java/com/amplitude/android/internal/locators/ComposeViewTargetLocator.kt
@file:Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE") // to access internal LayoutNode class

package com.amplitude.android.internal.locators

import ComposeLayoutNodeBoundsHelper
import androidx.compose.ui.node.LayoutNode
import androidx.compose.ui.node.Owner
import androidx.compose.ui.platform.InspectableValue
import com.amplitude.android.internal.ViewTarget
import com.amplitude.android.internal.compose.AmpFrustrationIgnoreElement
import com.amplitude.common.Logger
import java.util.ArrayDeque
import java.util.Queue

internal class ComposeViewTargetLocator(private val logger: Logger) : ViewTargetLocator {
    private val composeLayoutNodeBoundsHelper by lazy {
        ComposeLayoutNodeBoundsHelper(logger)
    }

    companion object {
        private const val SOURCE = "jetpack_compose"
    }

    override fun Any.locate(
        targetPosition: Pair<Float, Float>,
        targetType: ViewTarget.Type,
    ): ViewTarget? {
        val root = this as? Owner ?: return null

        val queue: Queue<LayoutNode> = ArrayDeque()
        queue.add(root.root)

        // the final tag to return (can be null if no test tag is found)
        var targetTag: String? = null

        // the final accessibility label to return
        var targetAccessibilityLabel: String? = null

        // track if we found a clickable element
        var foundClickableElement = false

        // the last known tag when iterating the node tree
        var lastKnownTag: String? = null

        // the last known accessibility label when iterating the node tree
        var lastKnownAccessibilityLabel: String? = null

        // Amplitude frustration analytics configuration (extracted as simple booleans)
        var ignoreRageClick = false
        var ignoreDeadClick = false

        while (queue.isNotEmpty()) {
            val node = queue.poll() ?: continue

            if (node.isPlaced &&
                composeLayoutNodeBoundsHelper.layoutNodeBoundsContain(node, targetPosition)
            ) {
                var isClickable = false
                val modifiers = node.getModifierInfo()

                for (modifierInfo in modifiers) {
                    val modifier = modifierInfo.modifier

                    // Check for Amplitude frustration analytics configuration
                    if (modifier is AmpFrustrationIgnoreElement) {
                        ignoreRageClick = modifier.ignoreRageClick
                        ignoreDeadClick = modifier.ignoreDeadClick
                    }

                    if (modifier is InspectableValue) {
                        when (modifier.nameFallback) {
                            "testTag" -> {
                                for (element in modifier.inspectableElements) {
                                    if (element.name == "tag") {
                                        lastKnownTag = element.value as? String
                                        break
                                    }
                                }
                            }

                            "semantics" -> {
                                for (element in modifier.inspectableElements) {
                                    if (element.name == "properties") {
                                        val elementValue = element.value
                                        if (elementValue is LinkedHashMap<*, *>) {
                                            for ((key, value) in elementValue.entries) {
                                                when (key) {
                                                    "TestTag" -> {
                                                        lastKnownTag = value as? String
                                                    }
                                                    "ContentDescription" -> {
                                                        // ContentDescription is a List<String> in Compose semantics
                                                        lastKnownAccessibilityLabel =
                                                            (value as? List<*>)
                                                                ?.filterIsInstance<String>()
                                                                ?.joinToString(", ")
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            }

                            "clickable" -> {
                                isClickable = true
                            }
                        }

                        val type = modifier.javaClass.name
                        if (type == "androidx.compose.foundation.ClickableElement" ||
                            type == "androidx.compose.foundation.CombinedClickableElement"
                        ) {
                            isClickable = true
                        }
                    }
                }

                if (isClickable && targetType == ViewTarget.Type.Clickable) {
                    foundClickableElement = true
                    targetTag = lastKnownTag // can be null if no test tag is found
                    targetAccessibilityLabel = lastKnownAccessibilityLabel // can be null
                }
            }
            queue.addAll(node.zSortedChildren.asMutableList())
        }

        return if (!foundClickableElement) {
            null
        } else {
            ViewTarget(
                _view = null,
                className = null,
                resourceName = null,
                tag = targetTag,
                text = null,
                accessibilityLabel = targetAccessibilityLabel,
                source = SOURCE,
                hierarchy = null,
                ampIgnoreRageClick = ignoreRageClick,
                ampIgnoreDeadClick = ignoreDeadClick,
            )
        }
    }
}
